gslib/utils/boto_util.py

```diff
@@ -387,6 +392,66 @@ def _PatchedGetPluginMethod(cls, requested_capability=None):
 
   boto.plugin.get_plugin = _PatchedGetPluginMethod
 
+  #########################################################################
+
+  # TODO(boto>2.49.0): Remove this.
+  # Fixes SSL issue where SNI was not being used for OpenSSL 1.1.1+
+  # https://github.com/boto/boto/pull/3843/files
+
+  # Import modules and resolve symbols used by our patch method.
+  import socket
+  import ssl
+  InvalidCertificateException = (
+      boto.https_connection.InvalidCertificateException)
+  ValidateCertificateHostname = (
+      boto.https_connection.ValidateCertificateHostname)
+
+  def _PatchedConnectMethod(self):
+    # The lines below were copied directly from the Boto file, so we don't lint
+    # or otherwise alter them.
+    if hasattr(self, "timeout"):
+        sock = socket.create_connection((self.host, self.port), self.timeout)
+    else:
+        sock = socket.create_connection((self.host, self.port))
+    msg = "wrapping ssl socket; "
+    if self.ca_certs:
+        msg += "CA certificate file=%s" % self.ca_certs
+    else:
+        msg += "using system provided SSL certs"
+    boto.log.debug(msg)
+    if hasattr(ssl, 'SSLContext') and getattr(ssl, 'HAS_SNI', False):
+        # Use SSLContext so we can specify server_hostname for SNI
+        # (Required for connections to storage.googleapis.com)
+        context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+        context.verify_mode = ssl.CERT_REQUIRED
+        if self.ca_certs:
+            context.load_verify_locations(self.ca_certs)
+        if self.cert_file:
+            context.load_cert_chain(self.cert_file, self.key_file)
+        self.sock = context.wrap_socket(sock, server_hostname=self.host)
+        # Add attributes only set in SSLSocket constructor without context:
+        self.sock.keyfile = self.key_file
+        self.sock.certfile = self.cert_file
+        self.sock.cert_reqs = context.verify_mode
+        self.sock.ssl_version = ssl.PROTOCOL_SSLv23
+        self.sock.ca_certs = self.ca_certs
+        self.sock.ciphers = None
+    else:
+        self.sock = ssl.wrap_socket(sock, keyfile=self.key_file,
+                                    certfile=self.cert_file,
+                                    cert_reqs=ssl.CERT_REQUIRED,
+                                    ca_certs=self.ca_certs)
+    cert = self.sock.getpeercert()
+    hostname = self.host.split(':', 0)[0]
+    if not ValidateCertificateHostname(cert, hostname):
+        raise InvalidCertificateException(hostname,
```

Comment:
Nitpick: is there any point we would want to print the __str__ method of this error? The only extra attribute you don't have in your print string is InvalidCertificateException.reason, and I'm not sure what that would contain that you don't already list. Might be ok for living behind the debug flag.

Reply:
Just copy/pasting what was added in the Boto PR.  This should suffice for most cases :)